warp10/src/main/java/io/warp10/script/functions/MOTIONSPLIT.java

```diff
@@ -32,174 +33,303 @@
+   * @param startLabel                  If set, add a label which will contain the start of the split
+   * @param timeThreshold               If the delay between two ticks goes beyond this value, force a split
+   * @param distanceThreshold           If we traveled more than distanceThreshold between two ticks, force a split
+   * @param proximityZoneRadius         Radius of the proximity zone in meters.
+   * @param proximityZoneTime           If we spent more than that much time in the proximity zone without moving and traveled slower than proximityInZoneMaxMeanSpeed while in the proximityZone force a split
+   * @param proximityZoneMaxSpeed       Maximum instant speed to consider you are in the same proximity zone
+   * @param proximityInZoneMaxMeanSpeed Minimum mean speed in the proximityZone to prevent a split, in m/s
+   * @param splitTypeLabel              If set, add a label with the reason of the split
+   * @param stoppedTimeLabel            If set, add a label with the time in the stop state
    * @return
    * @throws WarpScriptException
    */
-  private static List<GeoTimeSerie> motionSplit(GeoTimeSerie gts, String label, long timeThreshold, double distanceThreshold, double proximityZoneRadius, double proximityZoneSpeed, long proximityZoneTime) throws WarpScriptException {
+  private static List<GeoTimeSerie> motionSplit(GeoTimeSerie gts, String startLabel, long timeThreshold, double distanceThreshold, double proximityZoneRadius, double proximityZoneMaxSpeed,
+                                                long proximityZoneTime, double proximityInZoneMaxMeanSpeed, String splitTypeLabel, String stoppedTimeLabel) throws WarpScriptException {
     //
     // Sort GTS according to timestamps
     //
-    
+
     GTSHelper.sort(gts, false);
 
     GeoTimeSerie split = null;
-    
+
     List<GeoTimeSerie> splits = new ArrayList<GeoTimeSerie>();
-    
+
     int idx = 0;
-    
+
     int n = GTSHelper.nvalues(gts);
-    
+
     boolean mustSplit = true;
 
     long refLocation = GeoTimeSerie.NO_LOCATION;
     long refTimestamp = Long.MIN_VALUE;
-    
+
     long previousTimestamp = Long.MIN_VALUE;
     long previousLocation = GeoTimeSerie.NO_LOCATION;
     double proximityZoneTraveledDistance = 0.0D;
-    
+
     while (idx < n) {
       long timestamp = GTSHelper.tickAtIndex(gts, idx);
       long location = GTSHelper.locationAtIndex(gts, idx);
       long elevation = GTSHelper.elevationAtIndex(gts, idx);
       Object value = GTSHelper.valueAtIndex(gts, idx);
-    
+
+      //
+      // Initialise refLocation as soon as there is a valid position
+      //
+      if (GeoTimeSerie.NO_LOCATION == refLocation && GeoTimeSerie.NO_LOCATION != location) {
+        refLocation = location;
+        refTimestamp = timestamp;
+      }
+
+      //
+      // If the previous tick was more than 'timeThreshold' ago, split now
+      //
+      if (timestamp - previousTimestamp > timeThreshold) {
+        if (null != splitTypeLabel) {
```

Comment:
Simply set mustSplit to true and a splitReason to the reason for splitting, then set the label at the end of the loop.

Reply:
done